### PR TITLE
Align with Scala 3: Under `-Xsource:3`, remove by-name restriction for case copy

### LINKED ISSUE #7879: Wrong error message when using copy on case class with by-name parameter in second parameters list
```scala
/***
scalaVersion := "2.10.3-RC3"
 
*/
object Main extends App {
  case class UploadingFile(fid: String, name: String, length: Long, uploaded: Boolean)(onFileProcessingEnd: => String)
  val f = UploadingFile("", "", 0, true)("")
  println(f.copy(uploaded = true)())
 
}
```

```
[error] /tmp/rendererAQzTia6Gxe/src/main/scala/test.scala:8: value copy is not a member of Main.UploadingFile
[error]   println(f.copy(uploaded = true)())
while without by-nameness the error is more clear
[error] /tmp/rendererAQzTia6Gxe/src/main/scala/test.scala:8: not enough arguments for expression of type (onFileProcessingEnd: String)Main.UploadingFile.
[error] Unspecified value parameter onFileProcessingEnd.
[error]   println(f.copy(uploaded = true)())
```

btw why there is no default arguments now?
i've seen https://github.com/scala/scala-dist/pull/52/files but not sure what prompted that

### PR DESCRIPTION
**Pull request by som-snytt**
`-Xsource-features:case-copy-by-name` synthesizes case copy method when a parameter is by-name.

Fixes scala/bug#7879

### CODE REVIEW

**Comment by lrytz** [comment]
> putting behind `-Xsource-features`

I think we have to as it's changing the binary output.

File: src/compiler/scala/tools/nsc/typechecker/Unapplies.scala [hunk 1]
```diff
@@ -265,26 +265,24 @@ trait Unapplies extends ast.TreeDSL {
...
   def caseClassCopyMeth(cdef: ClassDef): Option[DefDef] = {
-    def isDisallowed(vd: ValDef) = isRepeatedParamType(vd.tpt) || isByNameParamType(vd.tpt)
-    val classParamss  = constrParamss(cdef)
-
-    if (cdef.symbol.hasAbstractFlag || mexists(classParamss)(isDisallowed)) None
-    else {
+    val classParamss = constrParamss(cdef)
+    def warn() = runReporting.warning(cdef.namePos, "case `copy` method is allowed to have by-name parameters under Scala 3 (or with -Xsource-features:case-copy-by-name)", Scala3Migration, cdef.symbol)
+    def isDisallowed(vd: ValDef) = isRepeatedParamType(vd.tpt) || isByNameParamType(vd.tpt).tap(if (_) warn())
+    def copyOK = currentRun.sourceFeatures.caseCopyByName || !mexists(classParamss)(isDisallowed)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: now it tries to generate `copy` for repeated params:

```
➜ sandbox scpr 10770 T.scala -Vprint:cleanup -Xsource:3-cross
> scala compile --server=false -S 2.13.15-bin-914c657-SNAPSHOT -d . -release 8 --repository https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots T.scala -Vprint:cleanup -Xsource:3-cross
/Users/luc/code/scala/scala13/sandbox/T.scala:1: error: type mismatch;
 found   : Seq[Int]
 required: Int
case class C(x: Int)(y: Int*)
                     ^
1 error
```

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: thanks, oops, I'll rename it to `sourceFeatures.anythingGoes`.

**Comment by som-snytt** [comment]
Refactored `copyOK` to use braces and `isAllowed` instead of `isDisallowed`. Added a neg test for varargs. (There is no existing test.)

File: src/compiler/scala/tools/nsc/typechecker/Unapplies.scala [hunk 2]
```diff
@@ -265,26 +265,29 @@ trait Unapplies extends ast.TreeDSL {
...
    */
   def caseClassCopyMeth(cdef: ClassDef): Option[DefDef] = {
-    def isDisallowed(vd: ValDef) = isRepeatedParamType(vd.tpt) || isByNameParamType(vd.tpt)
-    val classParamss  = constrParamss(cdef)
-
-    if (cdef.symbol.hasAbstractFlag || mexists(classParamss)(isDisallowed)) None
-    else {
+    val classParamss = constrParamss(cdef)
+    def copyOK = {
+      def warn() = runReporting.warning(cdef.namePos, "case `copy` method is allowed to have by-name parameters under Scala 3 (or with -Xsource-features:case-copy-by-name)", Scala3Migration, cdef.symbol)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]:
```suggestion
      def warn() = if (currentRun.isScala3) runReporting.warning(cdef.namePos, "case `copy` method is allowed to have by-name parameters under Scala 3 (or with -Xsource-features:case-copy-by-name)", Scala3Migration, cdef.symbol)
```

**Review by lrytz — CHANGES_REQUESTED** [review]
otherwise LGTM

File: src/compiler/scala/tools/nsc/typechecker/Unapplies.scala (hunk 2 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: added test

**Review by lrytz — APPROVED** [review]

**Pull request merged**